### ISSUE TITLE
Bug: ASO conversion webhook fails to properly convert resources applied by flux

### ISSUE THREAD

**Issue by ashetonh-ad** [Azure/azure-service-operator#3343]
**Version of Azure Service Operator**

```
❯ kubectl get deployment -n azureserviceoperator-system azureserviceoperator-controller-manager -o wide
NAME                                      READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS                IMAGES                                                                                         SELECTOR
azureserviceoperator-controller-manager   1/1     1            1           77d   manager,kube-rbac-proxy   mcr.microsoft.com/k8s/azureserviceoperator:v2.3.0,gcr.io/kubebuilder/kube-rbac-proxy:v0.13.1   control-plane=controller-manager
```

**Describe the bug**
After creating Sites and ServerFarms, flux tries to prune resources that randomly with this error

```
❯ k get kustomizations.kustomize.toolkit.fluxcd.io -n *****
NAME                  AGE   READY   STATUS
*******            41d   False   Site/***********/***-rest-na dry-run failed: failed to prune fields: failed add back owned items: failed to convert pruned object at version web.azure.com/v1api20220301: conversion webhook for web.azure.com/v1api20220301storage, Kind=Site returned invalid metadata: invalid metadata of type <nil> in input object...
```

**To Reproduce**
Haven't been able to isolate exactly what is going on but the steps that were taken to get here were:

- be on v2.1.0 and flux v2.0.0-rc.5 and kubernetes v1.26.3
- create a site and serverfarm via aso
- upgrade to v2.3.0 of ASO
- make changes to the site / serverfarm spec 
- after deleting the k8s resources (with the detach reconcile policy) and reconciling flux, it is able to work properly again (note that further down the line random other sites/serverfarms do the same thing, can't seem to find a pattern)

**Expected behavior**
I should expect that the ASO conversion webhook converts sites and serverfarms properly.

**Screenshots**

**Additional context**
Is there anything we can do around debugging the conversion webhook or making sure that the crd is the correct version for ASO v2.3.0?

**Comment by theunrepentantgeek** [comment]
Looking at this error message:
```
dry-run failed: failed to prune fields: failed add back owned items: 
failed to convert pruned object at version web.azure.com/v1api20220301: 
conversion webhook for web.azure.com/v1api20220301storage, Kind=Site returned invalid metadata: 
invalid metadata of type <nil> in input object...
```

There've been no changes to the [Site](https://azure.github.io/azure-service-operator/reference/web/) resource since v2.0.0, so I don't _think_ the problem is that your cluster has a stale (obsolete) CRD.

> Is there anything we can do around ... making sure that the crd is the correct version for ASO v2.3.0?

Restart the ASO container and monitor the logs as it starts up. You'll see log entries as it scans the cluster for known ASO CRDs and ensures they are up to date. If updates are needed, that should show up in the log, as will any errors.

```
dry-run failed: failed to prune fields: failed add back owned items: 
```

Do you know what Flux is trying to do here? Does it have expectations around the way resource-conversion works that are different from what api-server needs?

**Comment by matthchr** [comment]
This looks like it could be a duplicate of #3148 possibly?

**Comment by ashetonh-ad** [comment]
I made a new issue because I didn't think the premise held for our specific scenario. We aren't modifying anything, this is happening randomly.

**Comment by ashetonh-ad** [comment]
> Do you know what Flux is trying to do here? Does it have expectations around the way resource-conversion works that are different from what api-server needs?

That's just it. It shouldn't be trying to do anything because there is no diff, although somehow some fields must be changing downstream during conversion causing a diff to appear? I'm not entirely sure.

**Comment by theunrepentantgeek** [comment]
> Is there anything we can do around debugging the conversion webhook

You can pull logs from the ASO webhook using `kubectl`. If the conversion error comes from ASO, you should find the error logged in full. Please share it when you find it - the error you shared above is truncated, and I'm interested to see the gnarly details at the end.

_Aside_: Having the conversion webhooks fail was one of my nightmare scenarios during development; so much so that I code-generate `gopter` tests to exercise the code-generated conversions in order to try and catch any discrepancies. I'm super keen to work out what's going on here!

**Comment by matthchr** [comment]
Did some more searchign and I think this is due to this: https://github.com/kubernetes/kubernetes/issues/117356

**Comment by matthchr** [comment]
It does look like this is specifically related to server-side apply. I think you should be able to configure Flux to use client-side apply. If you do that, does the error stop happening?

**Comment by Shaked** [comment]
Hi, is there an ETA for a fix? We are experiencing this problem as well. 

Thank you

**Comment by matthchr** [comment]
I don't have an update here other than to say that we're waiting for the upstream fix. I'll take an action item to see how hard it might be for us to prod that along.

**Comment by matthchr** [comment]
There doesn't seem to be much movement upstream at the moment... I believe users have found workarounds and this issue doesn't strike that frequently. Unfortunately we're still waiting on the upstream fix.

**Comment by matthchr** [comment]
That upstream issue is still open but there seems to be a consensus (by lack of response) that post 1.29 this shouldn't be happening anymore.

I'm going to close this. We'll reopen if we see new reports.
